hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/TinyLfuBlockCache.java

```diff
@@ -196,13 +202,17 @@ public void cacheBlock(BlockCacheKey key, Cacheable value) {
             key.getHfileName(), key.getOffset(), value.heapSize(), DEFAULT_MAX_BLOCK_SIZE));
       }
     } else {
+      value.retain();
```

Comment:
We might have to have a closer look here.  Because now the reads from FS may be to shared buffers, we will have to close here.  See the LRU cache. I will also do a detailed read.

Reply:
We can make use of `isSharedMem()` similar to how LRU does it?
```
  private Cacheable asReferencedHeapBlock(Cacheable buf) {
    if (buf instanceof HFileBlock) {
      HFileBlock blk = ((HFileBlock) buf);
      if (blk.isSharedMem()) {
        return HFileBlock.deepCloneOnHeap(blk);
      }
    }
    // The block will be referenced by this LRUBlockCache, so should increase its refCnt here.
    return buf.retain();
  }
```
And instead of directly retaining value here, we can call this method. That seems like the only thing we are missing?